Test Framework (#312)
vdaas/vald: docs/developer/style.md

## Test

> Testing guideline has 3 important rules for the coding quality and readability

Reviewer: ```suggestion
There are 3 important things for the coding quality and readability.
```

### Table-Driven-Test

> Table-Driven-Test makes it easy to add new test case. Also it can avoid duplicating code. 

Reviewer: ```suggestion
Table-Driven-Test makes it easier to add new test case. Also it can avoid duplicating code. 
```

```

</td>
<td>

```go
test.New(
    test.WithCase(
        caser.New(
            WithName("case_1"),
            WithArg("192.0.2.0:8000")

>             WithWant("192.0.2.0", "8000", nil),

Reviewer: do we need `nil` here?

Author: Because `SplitHostPort` returns 3 objects (host, port, error), I think that `nil` is necessary.
`nil` is used by compares of error.

Reviewer: oh sorry I missed it. thank you for your answer :)

        ),
        caser.New(
            WithName("case_2"),
            WithArg("192.0.2.0:http")
            WithWant("192.0.2.0", "http", nil),
        ),
    ),
    test.WithDriverFunc(
        func(ctx context.Context, d test.DataProvider) []interface{} {
            host, port, err := net.SplitHostPort(d.Args()[0].(string))
            return []interface{} {
                host, port, err
            }
        }
    ),
).Run(context.Background(), t)
```

</td>

</tr>
</tbody>
</table>

> To create tests based on `Vald Style`, We provide the original test framework.

Reviewer: ```suggestion
To create tests based on `Vald Style`, we implement our testing framework.
```

> Please refer to [this page](../../internal/test).

Reviewer: ```suggestion
Please refer to the [test package](../../internal/test).
```
